tools/libSwiftScan/SwiftCaching.cpp

```diff
@@ -0,0 +1,902 @@
+  return false;
+}
+
+void swiftscan_cached_compilation_dispose(swiftscan_cached_compilation_t id) {
+  delete unwrap(id);
+}
+
+bool swiftscan_cached_output_load(swiftscan_cached_output_t output,
+                                  swiftscan_string_ref_t *error) {
+  auto *Out = unwrap(output);
+  // If proxy exists, there is nothing to be done.
+  if (Out->LoadedProxy)
+    return true;
+
+  auto failure = [&](llvm::Error E) {
+    *error =
+        swift::c_string_utils::create_clone(toString(std::move(E)).c_str());
+    return false;
+  };
+  if (auto Err = Out->CAS.getProxy(Out->Ref).moveInto(Out->LoadedProxy))
+    return failure(std::move(Err));
+
+  llvm::DenseSet<llvm::cas::ObjectRef> Visited;
+  llvm::SmallVector<llvm::cas::ObjectRef> WorkList;
+  auto addToWorkList = [&](llvm::cas::ObjectRef Item) {
+    if (Visited.insert(Item).second)
+      WorkList.push_back(Item);
+  };
+
+  addToWorkList(Out->LoadedProxy->getRef());
+  while (!WorkList.empty()) {
+    auto Current = WorkList.pop_back_val();
+    auto Proxy = Out->CAS.getProxyIfExists(Current);
+    if (!Proxy)
+      return failure(Proxy.takeError());
+
+    if (!*Proxy)
+      return false;
+
+    if (auto Err = (*Proxy)->forEachReference(
+            [&](llvm::cas::ObjectRef Ref) -> llvm::Error {
+              addToWorkList(Ref);
+              return llvm::Error::success();
+            }))
+      return failure(std::move(Err));
+  }
+  return true;
+}
+
+void swiftscan_cached_output_load_async(
+    swiftscan_cached_output_t output, void *ctx,
+    void (*callback)(void *ctx, bool success, swiftscan_string_ref_t error),
+    swiftscan_cache_cancellation_token_t *token) {
+  if (token)
+    *token = nullptr;
+
+  auto *Out = unwrap(output);
+  if (Out->LoadedProxy)
+    return callback(ctx, true, swift::c_string_utils::create_null());
+
+  /// Asynchronously visits the graph of the object node to ensure it's fully
+  /// materialized.
+  class AsyncObjectLoader
+      : public std::enable_shared_from_this<AsyncObjectLoader> {
+    void *Ctx;
+    void (*Callback)(void *Ctx, bool, swiftscan_string_ref_t);
+    SwiftCachedOutputHandle &Handle;
+
+    llvm::SmallDenseSet<llvm::cas::ObjectRef> ObjectsSeen;
+    unsigned NumPending = 0;
+    std::optional<llvm::cas::ObjectProxy> RootObj;
+    std::atomic<bool> MissingNode{false};
+    /// The first error that occurred.
+    std::optional<llvm::Error> ErrOccurred;
+    std::mutex Mutex;
+
+  public:
+    AsyncObjectLoader(void *Ctx,
+                      void (*Callback)(void *Ctx, bool, swiftscan_string_ref_t),
+                      SwiftCachedOutputHandle &Handle)
+        : Ctx(Ctx), Callback(Callback), Handle(Handle) {}
+
+    void visit(llvm::cas::ObjectRef Ref, bool IsRootNode) {
+      bool Inserted = false;
+      {
+        std::lock_guard<std::mutex> Guard(Mutex);
+        Inserted = ObjectsSeen.insert(Ref).second;
+        if (Inserted)
+          ++NumPending;
+      }
+      if (!Inserted) {
+        finishedNode();
```

Comment:
Oops, that is my mistake 😅but this line is wrong, it should be
```
      if (!Inserted) {
        return;
      }
```
without calling `finishedNode();`.

(I need to update the clang side..)